### DISCUSSION TITLE
Beta 0.0.2 Released!

### DISCUSSION THREAD

**Discussion by mertcandav** [julelang/jule#30] [Announcements]
# Beta 0.0.2 Release

This release focuses on further solidifying the compiler rather than adding new features. The first version released, Beta 0.0.1, was a version released after the bootstrapping phase. Beta 0.0.2, on the other hand, main purpose is to be more tested and bug-fixed, better optimized. Of course, this version brings with it a few new features.

Of course, this is not a stable release and may contain serious bugs. But Beta 0.0.1 was a rather buggy version. A more stable version therefore had to be developed quickly. Here is Beta 0.0.2 release. This release fixes many instability issues and fixes bugs. For the best Jule experience, we always recommend using the latest release.

---

> **Warning**
> The Beta 0.0.1 release is too buggy and outdated to compile current Jule codes.
> Therefore, if you are using it, it is strongly recommended that you upgrade to Beta 0.0.2.
> Otherwise you cannot use the compiler with the new standard library or compile from source.

## New Features

Beta 0.0.2 adds many experimental features. However, these features have been well tested and used for Beta 0.0.2 release, so they are not expected to cause serious problems.

### Compiler Optimizations

Compiler optimizations include some optimizations to make the generated object code more efficient and productive. Since this is the first release with optimizations added, it doesn't add any serious optimization support. But it adds dead code elimination which can significantly reduce your C++ compiler's compilation times, and copy optimization to reduce copies of runtime.

> Compiler optimizations were also used for Beta 0.0.2 builds.

Here is the manual page of [compiler optimizations](https://manual.jule.dev/compiler/compiler-optimizations.html).

### Cross Transpilation

The cross transpilation feature allows your compiler to generate IR code for different platforms, regardless of the platform you are on. This way you can create an IR and build on that platform your cross-platform apps, even if you don't have JuleC build for target platform.

> Cross transpilation were also used for create C++ IR of Beta 0.0.2 builds to get cross-platform builds.

Here is the manual page of [cross transpilation](https://manual.jule.dev/compiler/cross-transpilation.html).

### Capacity for Slices

This is more of an optimization than an addition. Slice now has a more capacity-oriented algorithm. For example, the old implementation of built-in `append` function allocates a new slice each time and performs copying. Now, if there is enough capacity, it adds to the destination slice and returns the slice itself, if there is not enough capacity, it allocates a new slice and copies the elements. This is an improvement that can significantly improve performance.

### Object Cleaning for Compile Mode

A simple cleaning process. When you use your compiler in build mode, the generated objects are left if you are using Beta 0.0.1 release. With Beta 0.0.2, your object directories are completely destroyed after compilation.

## Changes

- remove: string constructors

### JuleC

- compiler: set exit code to `1` for failed executions
- compiler: fix code generation for generic cpp-linked structs
- sema: fix built-in define importing of standard packages
- parser, sema: handle pointer type expressions
- sema: fix type declaration expression eval
- compiler: fix code generation of struct fields using cpp-linked struct
- sema: fix in-scope type alias processing
- sema: fix type alias eval
- sema: fix struct field checking
- compiler: fix struct ordering
- parser: fix generic type declaration parsing
- sema: fix struct type checking cycle occurs when struct has generics and uses itself in any method of itself
- parser: fix generic type parsing for grouped function return type declarations
- api: optimize reference counting atomicity
- parser: fix generic type declaration parsing for identifier types
- sema, compiler: fix octal escape sequence processing of strings
- sema: fix statically type checking
- sema: fixed automatic type annotation of assignments
- compiler: fix default value initialization of enum slice allocations
- sema: fix variadic argument eval
- lexer: fix carriage return character handling
- sema: fix struct field type checking
- sema: fix checking generic types of structures
- lexer: fix unicode processing
- compiler: fix code generation for sub-identifier accessor of common types
- sema: optimize auto-casting of expressions for assigned to `any` types

### API

- use: `static_cast` instead of `reinterpret_cast` as possible
- fix: `array.hpp` have not include directive for `<array>`
- add: `setup_envrionment_variables` function to `jule` namespace
- add: `environment_variables` global variable to `jule` namespace
- `jule::Trait<T>`: fix allocations
- fix: `Clone` derive's support for slices
- api: use referencing instead copying for deferred blocks
- minor optimizations

### Standard Library

- remove: `std::reflect`
- built-in `append` function: update algorithm and no longer takes immutable source argument
- std::vector: the `push` method of `Vector[T]` no longer takes variadic arguments
- std::vector: add `set`, `append`, `merge`, `merge_front`, `remove_range`, `insert`, and `slice` methods to `Vector[T]`
- std::jule::lexer: remove `parts`, `range_last`, and `range` functions
- std::jule::build: add `OS`, and `ARCH` globals
- std::os: add `ENV` global

**Comment by sparky4pro** [comment]
Thank you for this nice update.
